### ISSUE TITLE
refactoring needed

### ISSUE THREAD

**Issue by devkral** [mirumee/django-payments#138]
After slowly understanding the concepts, I see the problems in the code:
the payment providers are written by different people with different objectives (why somebody wants the payment provider to be able to change the name?, this concept can cause many headaches for others. And a create method of paypal will definitely break an automatic selection of payment providers).

Also there are some crossreferences on functions with unused parameters.

I will try to streamline the different mindsets for my project. The active branch in my fork is: payment_improvements

**Comment by devkral** [comment]
I don't want offend somebody. It just very hard to get into this code and I will try to refactor it

**Comment by Vesuvium** [comment]
I agree and I think that's a good idea, but if this is the case it might be interesting to write down a proper provider standard that can be used by anyone who wishes to contribute

**Comment by patrys** [comment]
We even consider splitting the main repository into the underlying API and separate gateway implementations (`django-payments-paypal` etc.).

**Comment by Vesuvium** [comment]
That would be great and would make lot of sense :) 

It would allow others to write payments providers without cluttering the main repo, as in my case where I have a regional provider and I had to use a fork.

**Comment by devkral** [comment]
look in my branch (in my fork) for progress. I am not ready with the paydirekt provider, as its refund capabilities are not tested

**Comment by devkral** [comment]
When I have sandbox access, I can complete the backend. Should be the next days.

**Comment by devkral** [comment]
Afterwards I will make a merge request. (it will be very big, as it contains some features and bugfixes)

**Comment by devkral** [comment]
I will add the backends later. See my pull requests; I splitted the changes into small pieces.
